### DISCUSSION TITLE
Cursor, selection, player: what do you think?

### DISCUSSION THREAD

**Discussion by guiv42** [helge17/tuxguitar#105] [General]
I propose a rework of the way the cursor is moved along when player plays/stops/loops, also considering click & drag selection feature introduced in 1.6.0.
Improvements are for sure possible. A few examples:
- click in measure 4, play, stop => cursor goes back to measure 4. Stop again => cursor goes back to measure 1 *(why?)*
- select measures 4 to 6 (left to right), type F9, check "play looped" then OK, then play => it starts playing at measure 6 *(seems counter-intuitive)*. And "stop" moves cursor back to measure 6 *(not better)*
- after selecting measures 4 to 6, it would be convenient that "play" action plays measures 4 to 6 once (or looped if "play looped" is selected)
- ...

But, experience has already shown users' expectations can conflict on this topic. And modifying this code without injecting side effects is also not trivial.

So, I would **really** like to have some consensus among users about the expected behavior **before** trying to implement anything. This is why I ask for your help: please give your opinion about the following questions!

# Cursor vs. selection
After selecting a few measures with click&drag, it would be nice if player starts from 1st selected measure when user presses "play" button, **whatever the direction of selection** (left to right, or right to left). Yet one role of cursor is to show graphically the *player starting point*. This is currently conflicting because cursor follows mouse, and its position at the end of selection process depends from selection direction. There I see 2 options:

1. when something is selected cursor is invisible. Implicitely, *player starting point* is graphically represented by the visible selected area
2. while selecting, cursor remains constantly aligned with selection start. It means it does not follow mouse any more if selection is done from left to right.

Intuitively I would think first option makes sense, but it creates an issue in the following scenario:

- select measures 4 to 6 (cursor is invisible)
- play => player starts at measure 4
- when in measure 5, pause => *player starting point* is now in measure 5. This needs to be shown.

In this case the only option I can see is to make cursor visible in measure 5. And it would not seem logical to clear selection. So in this situation, both selection and cursor shall be visible. If they can't be exclusive in this case, then I think it makes sense to keep both visible permanently.

So my proposal is to apply option 2.

_=> What is your opinion?_

I prototyped this behavior, and it looks pretty good to me.
 Here is what it would look like:

https://github.com/helge17/tuxguitar/assets/129443524/8ce89d9b-647d-4949-a1b6-8c349dcf732e

# Player mode dialog

Let's consider it becomes possible to play just an interval even if "play looped" is not checked in player mode dialog. This means the dialog shall be adapted. Basically, it should be permanently possible to define the "interval to play", with defaults set to "from beginning to end of song" to keep consistency with historical behavior.

Here is a prototype of such an adapted dialog (it's just a drawing made by hand: don't look at the detailed graphical layout, values, alignment of items, etc.).
The idea is that "Play from measure" and "Play to measure" fields can be manually entered in the dialog, or are directly updated when user selects an area with click&drag.

Note: when user updates "Play from measure" value, it would logically make sense to move cursor to that location after user has clicked OK.

_=> What is your opinion?_

![F9_proto](https://github.com/helge17/tuxguitar/assets/129443524/ba4b31d5-632d-4b74-857c-89e0d1e7b8a6)

**Comment by Sy256** [comment]
>Stop again => cursor goes back to measure 1 (why?)

I was not aware of this behavior but I do like it, it is sort of like how some DAWs work when you double click the stop button it jumps to the beginning. I have been using a shortcut to jump to the beginning instead, I dont mind if you change the behavior of the clicked buttons.

Hopefully I'm understanding these right let me know if i missed the point: On cursor vs selection I think if you have a span of more than 1 note place selected, the playback should start from the first selected note (not measure), and end at the last note of where the selection ends (unless looping is on). I would like if it could follow the selection span and not play outside it. So you could loop certain parts of measures instead of the whole thing for example. I dont believe I ever use the cursor separate than the selection, so I'm fine with the cursor (the little square box) going invisible when a selection is made. While the selection is made, playback should always start from the beginning of the selection. I rarely ever use the pause function (i hit space to play/stop) but if you PAUSE it in the loop the cursor could become visible again inside the selection, on the note it was paused on, and hitting pause/play again would start playback from that note. But stop/play would always reset the cursor to invisible and play from the beginning to end of the selection.

On the player mode dialog I am not sure about the cursor but I think the from and to should reflect whatever measures are selected by default. I would never deviate from the selection values but if someone wanted to, to start playback from the middle of a selection, and then continue playback from begin to end on loop… I guess the cursor could become visible on the new measure the same way it would if you pause/play it.

In the video the measures aren't as expanded as they would be with a bunch of notes, in actual practice for me I would probably lose the cursor within 3 measures if it's pinned to the beginning of the selections

**Comment by guiv42** [comment]
Thanks for this feedback. Several questions here.
1.
Where should the player start when first measure of selection is not fully selected?
It seems there is no consensus here. You suggest to start with the first selected note, and not at beginning of measure. But another user [mentioned](https://github.com/helge17/tuxguitar/issues/92#issuecomment-1694764228) he preferred to play the whole measure... In this case I have no plan to change that behavior

2.

> I'm fine with the cursor (the little square box) going invisible when a selection is made

In fact, this really creates an issue *from implementation point of view*. As I mentioned above, when pausing in the selection both selection and cursor shall definitely be shown. From software point of view I prefer to keep it simple: it's a lot easier to have either [cursor always visible], or [cursor visible only when nothing is selected]. It's really difficult to evaluate dynamically, depending from user's actions history (select, play, pause, stop, etc), if cursor should be visible or not.

3.
> On the player mode dialog I am not sure about the cursor but I think the from and to should reflect whatever measures are selected by default.

This is the way it currently works: when something is selected, if you open the player mode dialog then "from" and "to" values are filled from selection. No plan to change that.

4.
> In the video the measures aren't as expanded as they would be with a bunch of notes, in actual practice for me I would probably lose the cursor within 3 measures if it's pinned to the beginning of the selections

That's right, but is it really a problem? Currently, when several notes are selected, the only role of the cursor is to show the *player starting point*. I think you don't need in this case to see the cursor, as you would expect "play" action to start from selection's start. In fact it's not that different from [hiding cursor when selection is active], but it's a lot simpler to implement.

**Comment by Sy256** [comment]
Hey sorry I had work but to answer a bit of this: 

For 1: I agree with that other user but it sounds like they didn't specify about it when there's a span selected.Having a span of notes selected should play  within the span from the first note selected. But if no span is selected then it should start from the beginning of the measure of where the cursor is in as it currently does.

And I don't really care either way about the cursor when the selection is made _I think_ I was just saying if I made a selection span that wide and the cursor was stuck to the beginning I'd lose it anyway so it doesn't matter what it does, just the selection, so do whatever you think is best there

Thanks!

**Reply by guiv42** [reply]
OK that makes sense, thanks for your feedback

**Reply by Sy256** [reply]
ACTUALLY now that I think more about it I actually do prefer the cursor being pinned to the start of selection span, because it makes it easier to jump back to where the selection started by just hitting an arrow key. I would do this if I selected a bunch of measures to paste them to another track or something. Thank you!

**Comment by guiv42** [comment]
Significantly more complex than expected.
**Cursor vs. selection**: keeping cursor at the start of selection is **not possible**, it breaks the "Extend Selection" actions which can be very useful (especially if mapped to a shortcut). And for the same reason cursor needs to be visible even when a selection is present, it shows on which side the selection will be extended.
So it means cursor cannot always be used as a visual indicator for *player starting point*.

Currently working on another solution...

**Comment by guiv42** [comment]
see #139
